src/d3a/models/state.py

```diff
@@ -72,14 +73,50 @@ class ProsumptionInterface(StateInterface, ABC):
     def __init__(self):
         # Actual energy consumed/produced by the device at specific market slots
         self._energy_measurement_kWh: Dict[DateTime, float] = {}
+        self._unsettled_deviation_kWh: Dict[DateTime, float] = {}
+        self._forecast_measurement_deviation_kWh: Dict[DateTime, float] = {}
+
+    def _calculate_unsettled_energy_kWh(self, measured_energy_kWh, time_slot):
+        """Return the unsettled energy (produced or consumed) in kWh."""
+        return 0.0
 
     def set_energy_measurement_kWh(self, energy_kWh: float, time_slot: DateTime) -> None:
         """Set the actual energy consumed/produced by the device in the given market slot."""
         self._energy_measurement_kWh[time_slot] = energy_kWh
+        self._forecast_measurement_deviation_kWh[time_slot] = self._calculate_unsettled_energy_kWh(
+            energy_kWh, time_slot)
+        self._unsettled_deviation_kWh[time_slot] = \
+            abs(self._forecast_measurement_deviation_kWh[time_slot])
 
     def get_energy_measurement_kWh(self, time_slot: DateTime) -> float:
         """Get the actual energy consumed/produced by the device in the given market slot."""
-        return self._energy_measurement_kWh.get(time_slot, None)
+        return self._energy_measurement_kWh.get(time_slot)
+
+    def get_forecast_measurement_deviation_kWh(self, time_slot: DateTime) -> float:
+        """Get the energy deviation of forecasted energy from measurement by the device in
+        the given market slot. Negative value means that the deviation is beneficial to the
+        grid (and can be posted as an offer), positive value means that the deviation is
+        detrimental to the grid (and can be posted as a bid)"""
+        return self._forecast_measurement_deviation_kWh.get(time_slot)
+
+    def should_post_bid(self, time_slot: DateTime) -> bool:
+        return self._forecast_measurement_deviation_kWh.get(time_slot) > 0.0
+
+    def should_post_offer(self, time_slot: DateTime) -> bool:
+        return self._forecast_measurement_deviation_kWh.get(time_slot) < 0.0
+
+    def get_unsettled_deviation_kWh(self, time_slot: DateTime) -> float:
+        """Get the unsettled energy deviation of forecasted energy from measurement by the device
+        in the given market slot."""
+        return self._unsettled_deviation_kWh.get(time_slot)
+
+    def decrement_unsettled_deviation(
```

Comment:
I'm not sure whether this can be used by both production and consumption devices. If so, disregard this comment. If not, it should be moved into the respective class (either `ProductionState` or `ConsumptionState`).